File: examples/js-vuln-demo-no-answers/src/p1_safe_wrappers.js


class SafeQuery {
    static query(db, sql, ...params) {
        return db.query(sql, params);
    }
}


function findUser(db, userId) {
    return SafeQuery.query(db, 'SELECT * FROM users WHERE id = ?', userId);
}


function safeRead(base, name) {
    const path = require('path');
    const resolved = path.resolve(base, path.basename(name));
    if (!resolved.startsWith(path.resolve(base))) {
        throw new Error('Path traversal');
    }
    return require('fs').readFileSync(resolved, 'utf8');
}

File: examples/go-vuln-demo-no-answers/src/p1_safe_wrappers.go
package main

import (
	"database/sql"
	"path/filepath"
	"strings"
)


type SafeQuery struct{}

func (SafeQuery) Query(db *sql.DB, sqlStr string, params ...interface{}) (*sql.Row, error) {
	return db.QueryRow(sqlStr, params...)
}

var sq SafeQuery


func findUser(db *sql.DB, id int) (*sql.Row, error) {
	return sq.Query(db, "SELECT * FROM users WHERE id = ?", id)
}


func readConfig(base, name string) string {
	return filepathSafe(base, name)
}

func filepathSafe(base, name string) string {
	p := filepath.Join(base, name)
	if !strings.HasPrefix(p, filepath.Clean(base)) {
		return ""
	}
	return p
}
